rfd/0067-database-access-aws-redis

```diff
@@ -0,0 +1,260 @@
+generated, periodically rotated, and securely stored in the secret store (more
+on the secret store in a later section).
+
+In HA mode, if more than one database service tries to rotate the password for
+the same user at the same time, the secret store implementation guarantees only
+the first caller succeeds. And only the successor should call AWS APIs to
+update the database user's passwords.
+
+It has been found that it may take a few minutes for AWS to propagate user
+password changes to the Redis servers. To work around this, two versions of the
+passwords (`PREVIOUS` and `CURRENT`) are set to be effective AT THE SAME TIME
+for the database user, and `PREVIOUS` version can be used to guarantee success.
+For instance, a database user with passwords `v7` and `v8` is getting a
+password rotation to use `v8` and `v9`, after `v9` is created in the secret
+store. `v8` is a valid password while the password change is propagating or
+after. (From another point of view, `v8` is the current password and `v9` is
+the next.)
+
+When a client tries to connect to ElastiCache and MemoryDB Redis servers, the
+database service automatically logins Teleport managed database users by
+sending `auth` commands with the stored passwords to the Redis server. For
+database users not managed by Teleport, users can manually input `auth`.
+
+### Secret store
+
+#### Secret store - Interface
+
+For ElastiCache/MemoryDB Redis usage, it is natural to use an AWS native
+solution (aka [AWS Secrets Manager](https://aws.amazon.com/secrets-manager/)).
+Though other secret stores could be supported in the future. <sup>[2]<sup>
+
+Here is a sample interface for managing secrets using the secret store.
+```
+// SecretValue is the secret value.
+type SecretValue struct {
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret.
+    Value string
+    // Version is the version of the secret value.
+    Version string
+    // Created is the creation time of this version.
+    Created time.Time
+}
+
+// PutValueRequest defines the request to create a new secret version.
+type PutValueRequest struct {
+    // Key is the key path of the secret.
+    Key string
+    // Value is the value of the secret to set. A random value is generated if
+    // empty.
+    Value string
+    // RandomValueLength is the length of the random value to generate.
+    RandomValueLength int
+    // CurrentVersion is the version of the current secret value. If set, it
+    // will be used for test-and-set to ensure only one caller succeeds for the
+    // same version.
+    CurrentVersion string
+}
+
+// Secrets defines an interface for managing secrets. A secret consists of a
+// key path and a list of versions that hold copies of current or past secret
+// values.
+type Secrets interface {
+    // PutValue creates a new secret version for the secret. The secret is also
+    // created if it does not exist.
+    PutValue(req PutValueRequest) (*SecretValue, error)
```

Comment:
If the value already exist an error is returned?

Reply:
yes. i can add that to the description.